storage/watchable_store_test.go

```diff
@@ -59,6 +59,134 @@ func TestNewWatcherCancel(t *testing.T) {
+		synced: make(map[string]map[*watching]struct{}),
+	}
+
+	defer func() {
+		s.store.Close()
+		os.Remove(tmpPath)
+	}()
+
+	// Put a key so that we can spawn watchers on that key.
+	// (testKey in this test). This increases the rev to 1,
+	// and later we can we set the watcher's startRev to 1,
+	// and force watchers to be in unsynced.
+	testKey := []byte("foo")
+	testValue := []byte("bar")
+	s.Put(testKey, testValue)
+
+	w := s.NewWatcher()
+
+	// arbitrary number for watcher
+	watcherSize := 100
+
+	cancels := make([]CancelFunc, watcherSize)
+	for i := 0; i < watcherSize; i++ {
+		// use 1 to keep watchers in unsynced
+		_, cancel := w.Watch(testKey, true, 1)
+		cancels[i] = cancel
+	}
+
+	for idx := range cancels {
+		cancels[idx]()
+	}
+
+	// After running CancelFunc
+	//
+	// synced should still be empty
+	// because cancel does not move anything to synced
+	if len(s.synced[string(testKey)]) != 0 {
+		t.Errorf("synced[string(testKey)] size = %d, want 0", len(s.synced[string(testKey)]))
+	}
+	// unsynced should be empty
+	// because cancel deletes from unsynced
+	if len(s.unsynced) != 0 {
+		t.Errorf("unsynced size = %d, want 0", len(s.unsynced))
+	}
+}
+
+// TestSyncWatchings populates unsynced watching map and
+// tests syncWatchings method to see if it correctly sends
+// unsynced watchings to watchableStore.ch channel.
+func TestSyncWatchings(t *testing.T) {
+	s := &watchableStore{
+		store:    newStore(tmpPath),
+		unsynced: make(map[*watching]struct{}),
+		synced:   make(map[string]map[*watching]struct{}),
+	}
+
+	defer func() {
+		s.store.Close()
+		os.Remove(tmpPath)
+	}()
+
+	testKey := []byte("foo")
+	testValue := []byte("bar")
+	s.Put(testKey, testValue)
+
+	w := s.NewWatcher()
+
+	// arbitrary number for watcher
+	watcherSize := 100
+
+	for i := 0; i < watcherSize; i++ {
+		// use 1 to keep watchers in unsynced
+		w.Watch(testKey, true, 1)
+	}
+
+	// Before running s.syncWatchings()
+	//
+	// synced should be empty
+	// because we manually populate unsynced only
+	if len(s.synced[string(testKey)]) != 0 {
+		t.Fatalf("synced[string(testKey)] size = %d, want 0", len(s.synced[string(testKey)]))
+	}
+	// unsynced should not be empty
+	// because we manually populated unsynced only
+	if len(s.unsynced) == 0 {
+		t.Errorf("unsynced size = %d, want %d", len(s.unsynced), watcherSize)
+	}
+
+	// this should move all unsynced watchings
+	// to synced ones
+	s.syncWatchings()
+
+	// After running s.syncWatchings()
+	//
+	// synced should not be empty
+	// because syncWatchings populates synced
+	// in this test case
+	if len(s.synced[string(testKey)]) == 0 {
+		t.Errorf("synced[string(testKey)] size = 0, want %d", len(s.synced[string(testKey)]))
+	}
+	// unsynced should be empty
+	// because syncWatchings is expected to move
+	// all watchings from unsynced to synced
+	// in this test case
+	if len(s.unsynced) != 0 {
+		t.Errorf("unsynced size = %d, want 0", len(s.unsynced))
+	}
+
+	for w := range s.synced[string(testKey)] {
```

Comment:
how about just checking the ch size of the watcher that we created at line140? All of the watchings actually share one channel (https://github.com/coreos/etcd/blob/master/storage/watcher.go#L63), right?